src/util/focus-trap.ts

```diff
@@ -0,0 +1,140 @@
+import {DOCUMENT} from '@angular/common';
+import {Inject, Injectable, NgZone} from '@angular/core';
+
+const FOCUSABLE_ELEMENTS_SELECTOR = [
+  'a[href]', 'button:not([disabled])', 'input:not([disabled]):not([type="hidden"])', 'select:not([disabled])',
+  'textarea:not([disabled])', '[contenteditable]', '[tabindex]:not([tabindex="-1"])'
+].join(', ');
+
+enum DIRECTION {
+  BACKWARD,
+  FORWARD
+}
+
+/**
+ * Class that enforce the browser focus to be trapped inside a DOM element.
+ *
+ * The implementation is rather simple, the class add a `focusin` listener on the document with capture phase.
+ * Any focus event will then be caught, and therefore the class will only allow the one for elements contained inside
+ * it's own element.
+ *
+ * In case the element is not contained, the class will determine which new element has to be focused based on the `tab`
+ * navigation direction.
+ *
+ * Should not be used directly. Use only via {@link NgbFocusTrapFactory}
+ * @private
+ */
+export class NgbFocusTrap {
+  private _removeDocumentListener;
+  private _direction: DIRECTION = DIRECTION.FORWARD;
+  private _previouslyFocused: HTMLElement | null = null;
+
+  /**
+   * Register the focustrap instance to initially put the focus on specific element
+   *
+   * Simply add the `ngbAutofocus` attribute on the given element such as `<button ngbAutofocus></button>`
+   */
+  set autofocus(value: boolean) {
+    if (value === true && this._previouslyFocused === null) {
+      this._previouslyFocused = document.activeElement as HTMLElement;
+      this._focusInitial();
```

Comment:
This would be called synchronously, I'm not sure it will work for dynamically instantiated components. 
I guess it would try to focus before the change detection is run, so should be checked

Reply:
I am not sure it might be needed. Live tests seem to show that it is working in both datepicker and modal without that change.

Nonetheless, if we would ever need such a thing, I would rather prefer creating the focusTrap inside of an `ngAfterViewInit` to make sure that View is created and hence available to focus.